src/rgw/rgw_sal_filter.cc

```diff
@@ -0,0 +1,1362 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab ft=cpp
+
+/*
+ * Ceph - scalable distributed file system
+ *
+ * Copyright (C) 2022 Red Hat, Inc.
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation. See file COPYING.
+ *
+ */
+
+#include "rgw_sal_filter.h"
+
+namespace rgw { namespace sal {
+
+/* These are helpers for getting 'next' out of an object, handling nullptr */
+static inline PlacementTier* nextPlacementTier(PlacementTier* t)
+{
+  if (!t)
+    return nullptr;
+
+  return dynamic_cast<FilterPlacementTier*>(t)->get_next();
+}
+
+static inline User* nextUser(User* t)
+{
+  if (!t)
+    return nullptr;
+
+  return dynamic_cast<FilterUser*>(t)->get_next();
+}
+
+static inline Bucket* nextBucket(Bucket* t)
+{
+  if (!t)
+    return nullptr;
+
+  return dynamic_cast<FilterBucket*>(t)->get_next();
+}
+
+static inline Object* nextObject(Object* t)
+{
+  if (!t)
+    return nullptr;
+
+  return dynamic_cast<FilterObject*>(t)->get_next();
+}
+
+int FilterZoneGroup::get_placement_tier(const rgw_placement_rule& rule,
+					std::unique_ptr<PlacementTier>* tier)
+{
+  std::unique_ptr<PlacementTier> nt;
+  int ret;
+
+  ret = next->get_placement_tier(rule, &nt);
+  if (ret != 0)
+    return ret;
+
+  PlacementTier* t = new FilterPlacementTier(std::move(nt));
+  if (!t)
+    return -ENOMEM;
+
+  tier->reset(t);
+  return 0;
+}
+
+int FilterZone::get_zonegroup(const std::string& id,
+			      std::unique_ptr<ZoneGroup>* zonegroup)
+{
+  std::unique_ptr<ZoneGroup> ngz;
+  int ret;
+
+  ret = next->get_zonegroup(id, &ngz);
+  if (ret != 0)
+    return ret;
+
+  ZoneGroup* zg = new FilterZoneGroup(std::move(ngz));
+  if (!zg)
+    return -ENOMEM;
+
+  zonegroup->reset(zg);
+  return 0;
+}
+
+int FilterStore::initialize(CephContext *cct, const DoutPrefixProvider *dpp)
+{
+  zone = std::make_unique<FilterZone>(next->get_zone()->clone());
+
+  return 0;
+}
+
+const std::string FilterStore::get_name() const
+{
+  std::string name = "filter<" + next->get_name() + ">";
+  return name;
+}
+
+std::string FilterStore::get_cluster_id(const DoutPrefixProvider* dpp,  optional_yield y)
+{
+  return next->get_cluster_id(dpp, y);
+}
+
+std::unique_ptr<User> FilterStore::get_user(const rgw_user &u)
+{
+  std::unique_ptr<User> user = next->get_user(u);
+  return std::make_unique<FilterUser>(std::move(user));
+}
+
+int FilterStore::get_user_by_access_key(const DoutPrefixProvider* dpp, const std::string& key, optional_yield y, std::unique_ptr<User>* user)
+{
+  std::unique_ptr<User> nu;
+  int ret;
+
+  ret = next->get_user_by_access_key(dpp, key, y, &nu);
+  if (ret != 0)
+    return ret;
+
+  User* u = new FilterUser(std::move(nu));
+  if (!u)
+    return -ENOMEM;
+
+  user->reset(u);
```

Comment:
`new` won't return nullptr unless you use `new(std::nothrow) FilterUser`

let's not try to handle allocation failures here, just let any `std::bad_alloc` exceptions propagate

Reply:
Will fix.